components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/OrientCommands.java

```diff
@@ -53,31 +57,54 @@
 public class OrientCommands
     extends OConsoleDatabaseApp
 {
-  private static final String NL = System.getProperty("line.separator");
+  private static final Logger log = LoggerFactory.getLogger(OrientCommands.class);
 
   private static final String SCOPE = "orient";
 
+  private final BundleContext bundleContext;
+
+  // TODO: Consider a mass-command register/unregister component with mediator to avoid needing eager-singleton
+  // TODO: and injected bundle-context
+
   @Inject
-  public OrientCommands(final BundleContext ctx) {
+  public OrientCommands(final @Nullable BundleContext bundleContext) {
     super(new String[0]);
 
+    // HACK: should be non-null, but in test env is null
+    this.bundleContext = bundleContext;
+
     // define our own Karaf sub-shell: type 'orient' to enter, and 'exit' to leave
-    ctx.registerService(Function.class, createShellCommand(), config("*", SCOPE));
+    register(createShellCommand(), CommandHelper.config("*", SCOPE));
 
     for (Method method : getConsoleMethods().keySet()) {
       if (method.isAnnotationPresent(ConsoleCommand.class)) {
-        ctx.registerService(Function.class, createOrientCommand(method), config(SCOPE, method.getName()));
+        register(createOrientCommand(method), CommandHelper.config(SCOPE, method.getName()));
       }
     }
 
     onBefore(); // set OrientDB command defaults
   }
 
+  private void register(final Function function, final Dictionary<String,?> config) {
+    // HACK: should be non-null, but in test env is null
+    if (bundleContext == null) {
+      log.warn("Unable to register function, bundle-context is null: {}, config: {}", function, config);
+      return;
+    }
+
+    // TODO info -> debug
+    log.info("Registering function: {}, config: {}", function, config);
+    ServiceRegistration<Function> registration = bundleContext.registerService(Function.class, function, config);
+    // TODO: Handling unregistration?
+  }
+
   @Override
   protected void printApplicationInfo() {
     // hide banner as it doesn't apply here
   }
 
+  // TODO: Why is this not implementing Function, seems like unused Action support here mostly?
```

Comment:
It needs to implement Action to get the best integration with Karaf's shell, implementing Function is not enough